### DISCUSSION TITLE
Vim mode

### DISCUSSION THREAD

**Discussion by unrealapex** [monkeytypegame/monkeytype#1347] [Ideas] [answered]
It would be nice if there was a mode for Vim users. Navigation would be done using the `h` and `l` keys and invoking the command line would be done using `:`, `esc` would be used to go back to normal mode). Deleting a word would be done using `dw`, deleting a character would be done using the `x` key, and replacing a character would be done using the `r` key. The only challenge would be having a version of Monkeytype that was modal(normal, insert, visual, replace, etc...). What do you think?

**Comment by lukew3** [comment, marked as answer]
Imo, this feels out of the scope of the project. Keybindings in vim are useful for moving around a 2 dimensional document, and I don't see a need for it when typing in a single straight line like you do when taking a typing test.

**Reply by Miodec** [reply]
Yeah, this feels like a LOT of work for something that would be used by a very small group of people.

**Reply by unrealapex** [reply]
I must be the last of the almost extinct Vim using species. The IDEs are taking over. Now every program used to make code eats up your ram and hogs your cpu. Autocomplete, code refactoring, and debugging are now the standard for programming. What a world we have become lol.

**Reply by Rem0ld** [reply]
If I may, it could be great to implement the delete shortcut from vim, ctrl+h for one character and ctrl+w for one word.
I'm actually learning vim and it could be great to be able to train with those shortcuts while I'm training my typing speed.
And I don't think it should hard to make this as an option. May be I'm wrong though.

**Reply by Otard95** [reply]
To really make this feature worth it, it would have to allow for mode switching and moving in the typed text, movements, etc. etc.. as a vim user myself, though i see the appeal, i do still think this is out of scope.
Trust me when i say `vimtutor` is the better option.

**Reply by unrealapex** [reply]
> ctrl+h for one character and ctrl+w for one word.

I'm pretty sure it's `x` for one character and `dw` for words, but I do understand your reasoning.

**Reply by Rem0ld** [reply]
> > ctrl+h for one character and ctrl+w for one word.
> 
> I'm pretty sure it's `x` for one character and `dw` for words, but I do understand your reasoning.

While in insert mode, you can ctrl+h for one letter ctrl+w for one word ctrl+u for one sentence, so you don't have to return to normal mode to edit and you don't have to reach backspace to delete 😉. Even better if you mapped ctrl at the caps lock key

**Reply by Otard95** [reply]
> > I'm pretty sure it's `x` for one character and `dw` for words, but I do understand your reasoning.
> 
> While in insert mode, you can ctrl+h for one letter ctrl+w for one word ctrl+u for one sentence, so you don't have to return to normal mode to edit and you don't have to reach escape to delete 😉. Even better if you mapped ctrl at the caps lock key

`ctrl+backspace` also deletes the preceding word 👍

**Reply by Rem0ld** [reply]
> > > I'm pretty sure it's `x` for one character and `dw` for words, but I do understand your reasoning.
> > 
> > 
> > While in insert mode, you can ctrl+h for one letter ctrl+w for one word ctrl+u for one sentence, so you don't have to return to normal mode to edit and you don't have to reach escape to delete 😉. Even better if you mapped ctrl at the caps lock key
> 
> `ctrl+backspace` also deletes the preceding word 👍

The idea is not to have to use backspace^^ I know you can delete word and sentence with shortcuts (I'm using them)

**Reply by unrealapex** [reply]
> Even better if you mapped ctrl at the caps lock key

I suggest binding escape to caps lock, makes going back to normal mode a whole lot easier :)

**Reply by unrealapex** [reply]
> > > ctrl+h for one character and ctrl+w for one word.
> > 
> > 
> > I'm pretty sure it's `x` for one character and `dw` for words, but I do understand your reasoning.
> 
> While in insert mode, you can ctrl+h for one letter ctrl+w for one word ctrl+u for one sentence, so you don't have to return to normal mode to edit and you don't have to reach backspace to delete 😉. Even better if you mapped ctrl at the caps lock key

Didn't know that, I'll try it 😃 .

**Reply by azinsharaf** [reply]
Hi, as a Neovim user,  i see ctrl-h (delete a letter) works in monkeytype but ctrl-w doesn't. Is it right?

**Reply by unrealapex** [reply]
I would assume it depends on your browser's behavior for the keyboard shortcut. Certain browsers have this key mapped such as <kbd>CTRL + H</kbd> in Google Chrome for opening Chrome's history.

**Reply by azinsharaf** [reply]
same experience on both Brave and Safari. Does ctrl-w work for you?

**Reply by Rem0ld** [reply]
I'm a mac user and cmd-w only work on some apps (mainly what we could call, dev apps) but cmd-h works everywhere

**Reply by unrealapex** [reply]
<kbd>CTRL + W</kbd> does not work for me(Chrome). I will check on other browsers.

**Reply by azinsharaf** [reply]
Is `crrl-w` supposed to work in monkeytype?

**Reply by Rem0ld** [reply]
Nope

**Reply by azinsharaf** [reply]
Vim is not used by a very small amount of people (it was among top 5  IDE on 2022 in SO survey). The idea of using the vim mode in monkey type doesn't have anything do with learning vim. Vim users already use ctrl-w so when they are learning typing using this great platform they like to use the shortcuts they already know. And ctrl-w is the important one. I agree there is no need for hjkl here since typing is in insert mode and ctrl-w is the most efficient way for a vim user to delete a word. 
I don't know how much effort is needed for this implementation so i am gonna add ctrl-back space or caps-lock-backspace to my muscle memory for now.
